src/ucp/wireup/wireup_cm.c

```diff
@@ -0,0 +1,272 @@
+
+
+static unsigned
+ucp_cm_ep_init_flags(const ucp_worker_h worker, const ucp_ep_params_t *params)
+{
+    if (!ucp_worker_sockaddr_is_cm_proto(worker)) {
+        return 0;
+    }
+
+    if (params->field_mask & UCP_EP_PARAM_FIELD_SOCK_ADDR) {
+        return UCP_EP_INIT_CM_WIREUP_CLIENT;
+    }
+
+    if (params->field_mask & UCP_EP_PARAM_FIELD_CONN_REQUEST) {
+        return UCP_EP_INIT_CM_WIREUP_SERVER;
+    }
+
+    return 0;
+}
+
+static ucs_status_t ucp_cm_preinit_client_lanes(ucp_ep_h ucp_ep)
+{
+    ucp_worker_h worker        = ucp_ep->worker;
+    ucp_ep_config_key_t key    = ucp_ep_config(ucp_ep)->key;
+    uint64_t addr_pack_flags   = UCP_ADDRESS_PACK_FLAG_DEVICE_ADDR |
+                                 UCP_ADDRESS_PACK_FLAG_IFACE_ADDR  |
+                                 /* Disable HW AMO because ucp_ep lanes are
+                                  * bound to specific device which may be
+                                  * different form worker->atomic_tls */
+                                 UCP_ADDRESS_PACK_FLAG_DISABLE_HW_AMO;
+    ucp_wireup_ep_t *wireup_ep = ucp_ep_get_cm_wireup_ep(ucp_ep);
+    void *ucp_addr;
+    size_t ucp_addr_size;
+    ucp_unpacked_address_t unpacked_addr;
+    uint8_t *addr_indices;
+    ucp_lane_index_t lane_idx;
+    ucs_status_t status;
+
+    ucs_assert_always(wireup_ep != NULL);
+
+    /* Construct local dummy address for lanes selection taking an assumption
+     * that server has the transports which are the best from client's
+     * perspective. */
+    status = ucp_address_pack(worker, NULL,
+                              ucp_context_tl_bitmap(worker->context,
+                                                    wireup_ep->dev_name),
+                              addr_pack_flags, NULL, &ucp_addr_size, &ucp_addr);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    status = ucp_address_unpack(worker, ucp_addr, addr_pack_flags,
+                                &unpacked_addr);
+    if (status != UCS_OK) {
+        goto free_ucp_addr;
+    }
+
+    addr_indices = ucs_malloc(unpacked_addr.address_count *
+                              sizeof(*addr_indices), "dummy addr indices");
+    if (addr_indices == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto free_addr_list;
+    }
+
+    status = ucp_wireup_select_lanes(ucp_ep, &wireup_ep->ucp_ep_params,
+                                     ucp_cm_ep_init_flags(worker,
+                                                          &wireup_ep->ucp_ep_params),
+                                     unpacked_addr.address_count,
+                                     unpacked_addr.address_list, addr_indices,
+                                     &key);
+    if (status != UCS_OK) {
+        goto free_addr_indicies;
+    }
+
+    status = ucp_worker_get_ep_config(worker, &key, 0, &ucp_ep->cfg_index);
+    if (status != UCS_OK) {
+        goto free_addr_indicies;
+    }
+
+    ucp_ep->am_lane = key.am_lane;
+
+    /* 0-lane is a CM lane */
```

Comment:
ucs_assert this

Reply:
this code is gone in #4283 after merging 2 loops thru lanes into single one